src/lib/support/StringCircularBuffer.cpp

```diff
@@ -0,0 +1,151 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <support/StringCircularBuffer.h>
+
+#include <limits>
+#include <nlassert.h>
+#include <string.h>
+
+#include <support/CodeUtils.h>
+
+namespace chip {
+
+std::size_t StringCircularBuffer::Advance(std::size_t dataLocation, size_t amount) const
+{
+    dataLocation += amount;
+    if (dataLocation >= mCapacity)
+        dataLocation -= mCapacity;
```

Comment:
This does not behave very well when `amount + dataLocation > 2*mCapacity`...  We should document the assumption of this method that `amount <= mCapacity` and then double-check that all callers enforce this.

Reply:
```cpp
return (dataLocation + amount) % mCapacity;
```

seems to be better behaved. Is it the same? Do we really need a separate method if this is all that is needed?

Reply:
The point is if either `amount` or `dataLocation` is `>= mCapacity` then something is really broken and this method can't do its work at all.  It just needs to document its expected invariants....

Using `%` would end up with a value inside the buffer, yes, but the value would still be meaningless in practice.

Reply:
Added a pre-condition that `amount` must be less than `mCapacity`, also verified that all calling site meet the condition.